Assets/MixedRealityToolkit.SDK/Experimental/Features/Dwell/DwellHandler.cs

```diff
@@ -0,0 +1,243 @@
+                    case DwellStateType.FocusGained:
+                        return 0;
+                    case DwellStateType.DwellStarted:
+                        return Mathf.Clamp((float)(dwellProfile.TimeToCompleteDwell - FillTimer) / dwellProfile.TimeToCompleteDwell, 0f, 1f);
+                    case DwellStateType.DwellCompleted:
+                        return 1;
+                    case DwellStateType.DwellCanceled:
+                        if ((DateTime.UtcNow - focusExitTime).TotalSeconds > dwellProfile.TimeToAllowDwellResume)
+                        {
+                            CurrentDwellState = DwellStateType.None;
+                            return 0;
+                        }
+                        else if (dwellProfile.TimeToAllowDwellResume > 0)
+                        {
+                            return Mathf.Clamp((float)(dwellProfile.TimeToCompleteDwell - FillTimer) / dwellProfile.TimeToCompleteDwell, 0f, 1f);
+                        }
+                        break;
+                    default:
+                        return 0;
+                }
+
+                return 0;
+            }
+        }
+
+        /// <summary>
+        /// Cached pointer reference to track focus events maps to teh same pointer that initiated dwell
+        /// </summary>
+        private IMixedRealityPointer pointer;
+
+        private DateTime focusEnterTime = DateTime.MaxValue;
+        private DateTime focusExitTime = DateTime.MaxValue;
+
+        /// <summary>
+        /// Exposes whether the target has focus from the pointer defined in dwell profile settings
+        /// </summary>
+        protected bool HasFocus { get; private set; }
+
+        /// <summary>
+        /// Captures the dwell status 
+        /// </summary>
+        protected DwellStateType CurrentDwellState = DwellStateType.None;
+
+        /// <summary>
+        /// Abstracted value for the how long the dwelled object still needs to be focused to comlete the dwell action
+        /// </summary>
+        protected float FillTimer = 0;
+
+        private void Awake()
+        {
+            Debug.Assert(dwellProfile != null, "DwellProfile is null, creating default profile.");
+
+            if (dwellProfile == null)
+            {
+                dwellProfile = ScriptableObject.CreateInstance<DwellProfile>();
+            }
+
+            CancelDwellAction += OnDwellCancelAction;
+        }
+
+        private void Update()
+        {
+            UpdateFillTimer();
+
+            if (HasFocus && CurrentDwellState != DwellStateType.DwellCompleted)
+            {
+                double focusDuration = (DateTime.UtcNow - this.focusEnterTime).TotalSeconds;
+
+                if (CurrentDwellState == DwellStateType.FocusGained && focusDuration >= dwellProfile.DwellIntentDelay)
+                {
+                    DwellIntended.Invoke(pointer);
+                    CurrentDwellState = DwellStateType.DwellIntended;
```

Comment:
Does the order matter here? If the caller asks for the current state, they will still see that it's in the FocusGained state.

Or is that the intent (in which case having a comment saying that this is the general pattern would also help to provide a better understanding of the mechanics of how these things are used)

Reply:
What I mean by this is, I'm assuming DwellIntended gets invoked sync here, which means if the caller tries to ask this dwell handler what state its in (because maybe it's calling some common code that is not aware it's being called from DwellIntended) this could be weird maybe

---
In reply to: [314431451](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5594#discussion_r314431451) [](ancestors = 314431451)

Reply:
The CurrentDwellState  is a protected variable and so the caller does not have access to its value. But i think it is good to invoke the events after state transitions for consistent pattern.